recvattach.c

```diff
@@ -456,36 +456,26 @@ int attach_tag(struct Menu *menu, int sel, int act)
 
 /**
  * prepend_savedir - Add #C_AttachSaveDir to the beginning of a path
- * @param buf     Buffer for the result, must be valid
- * @param bufsize Size of the buffer
+ * @param buf Buffer for the result
  */
-static void prepend_savedir(char *buf, size_t bufsize)
+static void prepend_savedir(struct Buffer *buf)
 {
-  const char *savedir = C_AttachSaveDir;
-
-  if (buf[0] == '/')
+  if (!buf || !buf->data || (buf->data[0] == '/'))
     return;
 
-  if (!savedir || !*savedir)
-    savedir = "./";
-
-  size_t savedirlen = strlen(savedir);
-  size_t buflen = strlen(buf);
-  bool addsep = (savedir[savedirlen - 1] != '/');
-  size_t newbuflen = savedirlen + buflen + addsep;
-
-  if (bufsize < newbuflen)
+  struct Buffer *tmp = mutt_buffer_pool_get();
+  if (C_AttachSaveDir)
   {
-    return;
+    mutt_buffer_addstr(tmp, C_AttachSaveDir);
+    if (tmp->dptr[-1] != '/')
```

Comment:
I've never seen `-1` used as an index. What does it do?

Reply:
It's simply the previous character, the one before `tmp->dptr[0]`

In this context, it's a little bit hacky.
The `Buffer` holds a large array of `char`, and `dptr` keeps track of the `NUL` at the end.
A string appended to the `Buffer` will start at `dptr`.
`dptr[0] == '\0'`
`dptr[-1]` is the last character of the string in the `Buffer` (which we know isn't empty).

Reply:
> It's simply the previous character, the one before tmp->dptr[0]

Isn't that undefined behavior since we're accessing data outside the bounds of `dptr`? 

I understand that it's safe here because we know `dptr` points to the `Buffer`'s `char`, I'm just curious.

Reply:
No.  It would only be undefined behaviour if we went outside of the block of memory that the `Buffer` allocated.
Within that block, we're free to reference it any way we choose.

> I understand that it's safe here

Yeah and that's quite a few assumptions.
We _ought_ to check that:
- `data != NULL` (there's data)
- `dsize != 0` (ditto)
- `dptr < (data + dsize - 1)` (ptr is within range)